diff --git a/Makefile b/Makefile
--- a/Makefile  2013-06-26 20:05:20.000000000 +0800
+++ b/Makefile  2013-11-19 22:39:45.000000000 +0800
@@ -33,11 +33,11 @@
 LD_SET_SONAME = -Wl,-soname=
 INSTALL_FLAGS = -D -m

-LDSO_PATHNAME = libproxychains4.$(LDSO_SUFFIX)
+LDSO_PATHNAME = libproxychains.$(LDSO_SUFFIX)

 SHARED_LIBS = $(LDSO_PATHNAME)
 ALL_LIBS = $(SHARED_LIBS)
-PXCHAINS = proxychains4
+PXCHAINS = proxychains
 ALL_TOOLS = $(PXCHAINS)

 -include config.mak
