--- fcitx-4.2.9.1/src/lib/fcitx-gclient/CMakeLists.txt	2017-02-06 16:01:13.532123698 +0800
+++ fcitx-4.2.9.1.new/src/lib/fcitx-gclient/CMakeLists.txt	2017-02-06 16:03:19.178121134 +0800
@@ -86,7 +86,8 @@
 endif()

 if(ENABLE_STATIC)
-  add_library(fcitx-gclient-static STATIC ${FCITX_GCLIENT_SOURCES})
+  add_library(fcitx-gclient-static STATIC ${FCITX_GCLIENT_SOURCES}
+  ${FCITX_GCLIENT_BUILT_SOURCES} ${FCITX_GCLIENT_BUILT_HEADERS})
   set_target_properties(fcitx-gclient-static PROPERTIES
     OUTPUT_NAME "fcitx-gclient"
     PREFIX "lib")
