diff -ru a/htop.c b/htop.c
--- a/htop.c	2014-11-11 17:14:50.372661676 +0800
+++ b/htop.c	2014-11-11 17:13:25.133659574 +0800
@@ -70,7 +70,7 @@
 }

 static struct { const char* key; const char* info; } helpLeft[] = {
-   { .key = " Arrows: ", .info = "scroll process list" },
+   { .key = "   hjkl: ", .info = "scroll process list" },
    { .key = " Digits: ", .info = "incremental PID search" },
    { .key = "   F3 /: ", .info = "incremental name search" },
    { .key = "   F4 \\: ",.info = "incremental name filtering" },
@@ -90,7 +90,7 @@
    { .key = "  Space: ", .info = "tag process" },
    { .key = "      c: ", .info = "tag process and its children" },
    { .key = "      U: ", .info = "untag all processes" },
-   { .key = "   F9 k: ", .info = "kill process/tagged processes" },
+   { .key = "   F9 d: ", .info = "kill process/tagged processes" },
    { .key = "   F7 ]: ", .info = "higher priority (root only)" },
    { .key = "   F8 [: ", .info = "lower priority (+ nice)" },
 #if (HAVE_LIBHWLOC || HAVE_NATIVE_AFFINITY)
@@ -101,7 +101,7 @@
    { .key = "      s: ", .info = "trace syscalls with strace" },
    { .key = "         ", .info = "" },
    { .key = "   F2 S: ", .info = "setup" },
-   { .key = "   F1 h: ", .info = "show this help screen" },
+   { .key = "   F1 ?: ", .info = "show this help screen" },
    { .key = "  F10 q: ", .info = "quit" },
    { .key = NULL, .info = NULL }
 };
@@ -684,7 +684,6 @@
          break;
       }
       case KEY_F(1):
-      case 'h':
       case '?':
       {
          showHelp(pl);
@@ -720,7 +719,7 @@
          CRT_enableDelay();
          break;
       }
-      case 'l':
+      case 'L':
       {
          Process* p = (Process*) Panel_getSelected(panel);
          if (!p) break;
@@ -783,7 +782,7 @@
          break;
       }
       case KEY_F(9):
-      case 'k':
+      case 'd':
       {
          Panel* signalsPanel = (Panel*) SignalsPanel_new();
          const char* fuFunctions[] = {"Send  ", "Cancel ", NULL};
diff -ru htop-1.0.3/Panel.c a/Panel.c
--- a/Panel.c	2014-11-11 17:14:55.957661814 +0800
+++ b/Panel.c	2014-11-11 17:14:20.958660951 +0800
@@ -343,11 +343,13 @@
    assert (this != NULL);
    switch (key) {
    case KEY_DOWN:
+   case 'j':
    case KEY_CTRLN:
       if (this->selected + 1 < Vector_size(this->items))
          this->selected++;
       return true;
    case KEY_UP:
+   case 'k':
    case KEY_CTRLP:
       if (this->selected > 0)
          this->selected--;
@@ -375,6 +377,7 @@
       return true;
    #endif
    case KEY_LEFT:
+   case 'h':
    case KEY_CTRLB:
       if (this->scrollH > 0) {
          this->scrollH -= CRT_scrollHAmount;
@@ -382,6 +385,7 @@
       }
       return true;
    case KEY_RIGHT:
+   case 'l':
    case KEY_CTRLF:
       this->scrollH += CRT_scrollHAmount;
       this->needsRedraw = true;
