--- src/sunstone/OpenNebulaVNC.rb	2014-06-11 21:36:16.000000000 +0800
+++ src/sunstone/OpenNebulaVNC.rb	2014-07-06 19:55:43.393463973 +0800
@@ -119,7 +119,7 @@
             proxy_options << " -6"
         end

-        cmd ="python #{@proxy_path} #{proxy_options} #{@proxy_port}"
+        cmd ="python2 #{@proxy_path} #{proxy_options} #{@proxy_port}"

         begin
             @logger.info { "Starting VNC proxy: #{cmd}" }
